### ISSUE TITLE
TensorOptions extensibility has rusted shut

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#42863]
One of the original motivations for introducing TensorOptions was to make it easier to add new parameters to factory functions, should the need arise, without having to modify all function signatures; it'd just be a new field in TensorOptions. Furthermore, as TensorOptions were always kwarg-only arguments on Python side, there wouldn't be any BC implications for changes like this.

With the addition of mobile serialization that is position dependent, we no longer can rely on an argument being kwarg-only; position matters, no matter what.  However, this doesn't necessarily imply that we can't add new parameters to TensorOptions; if the newly added parameters are defaulted, the new definitions are "roughly" backwards compatible.

However, we also now have native functions which add new kwargs AFTER the tensor options kwargs:

```
- func: empty.memory_format(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor
```

For these functions, TensorOptions extensibility is no longer a thing. TensorOptions parameters should live in a contiguous segment `ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None`. But to insert a new parameter between `pin_memory` and `memory_format` is BC-breaking as it changes the positional order of arguments. So we have no way to add new arguments now. :(

Note: Can we just add the new argument at the very end of the argument list? Maybe. It would be pretty strange though: when you translate back into the TensorOptions version of the code, arguably you should delete the new argument from the argument list; so now you are in the weird situation where `pin_memory` is not the last argument in schema, but it is the last argument in C++ API.

**Comment by bhosmer** [comment]
What if we relaxed our rule so that the scattered arguments could be appear anywhere in the signature - out of order, discontiguous - as long as the names (or names and types?) were recognized as TensorOptions constituents?

**Comment by ezyang** [comment]
I thought about this too. It sounds kind of complicated though; certainly all of our current codegen would break.

**Comment by bhosmer** [comment]
True, it would definitely need an upgrade to the current codegen. But upgrading the pattern matching logic itself would be pretty straightforward - probably the most arduous part would be digging out all the places that assume the current invariant, but I think it's doable.
